three digit date 225 gives exception org.nineml.coffeegrinder.exceptions.TreeWalkerException: Unknown error: T005

I have returned to try to use Coffeesacks and that way Coffeegrinder, when using the date.ixml of
```
date: s?, day, -s, month, (-s, year)? .
-s: -" "+ .
day: digit, digit? .
-digit: "0"; "1"; "2"; "3"; "4"; "5"; "6"; "7"; "8"; "9".
month: "January"; "February"; "March"; "April";
       "May"; "June"; "July"; "August";
       "September"; "October"; "November"; "December".
year: (digit, digit)?, digit, digit .
```
with the XSLT 3.0 stylesheet
```xslt
<?xml version="1.0" encoding="utf-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
                xmlns:cs="http://nineml.com/ns/coffeesacks"
                exclude-result-prefixes="#all"
                expand-text="yes"
                version="3.0">

    <xsl:output method="xml" encoding="utf-8" indent="yes"/>

    <xsl:template match="/" name="xsl:initial-template">
        <xsl:variable name="parser" select="cs:load-grammar('date.ixml')"/>
        <date>
            <xsl:sequence select="$parser('05 February 225')"/>
        </date>
    </xsl:template>

</xsl:stylesheet>
```
I hoped to get some error that parsing failed due to a missing digit for the date but instead I get a longish stack trace:
```
Error on line 12 column 15 of sheet2.xsl:
   org.nineml.coffeegrinder.exceptions.TreeWalkerException: Unknown error: T005
org.nineml.coffeegrinder.exceptions.TreeWalkerException: Unknown error: T005
	at org.nineml.coffeegrinder.exceptions.TreeWalkerException.getException(TreeWalkerException.java:30)
	at org.nineml.coffeegrinder.exceptions.TreeWalkerException.noTrees(TreeWalkerException.java:64)
	at org.nineml.coffeegrinder.trees.Lumberjack.getTree(Lumberjack.java:72)
	at org.nineml.coffeesacks.CommonDefinition$InvisibleXmlParserFunction.call(CommonDefinition.java:502)
	at net.sf.saxon.functions.CallableFunction.call(CallableFunction.java:128)
	at net.sf.saxon.expr.DynamicFunctionCall$DynamicFunctionCallElaborator.lambda$elaborateForPull$2(DynamicFunctionCall.java:359)
	at net.sf.saxon.expr.elab.PullElaborator.lambda$elaborateForPush$0(PullElaborator.java:38)
	at net.sf.saxon.expr.instruct.FixedElement$FixedElementElaborator.lambda$elaborateForPush$0(FixedElement.java:640)
	at net.sf.saxon.expr.instruct.NamedTemplate.expand(NamedTemplate.java:247)
	at net.sf.saxon.trans.XsltController.callTemplate(XsltController.java:872)
	at net.sf.saxon.s9api.Xslt30Transformer.callTemplate(Xslt30Transformer.java:503)
	at org.example.Main.main(Main.java:34)
: Unknown error: T005
In template xsl:initial-template on line 10 column 57 of sheet2.xsl:
org.nineml.coffeegrinder.exceptions.TreeWalkerException: Unknown error: T005
net.sf.saxon.s9api.SaxonApiException: org.nineml.coffeegrinder.exceptions.TreeWalkerException: Unknown error: T005
	at net.sf.saxon.s9api.Xslt30Transformer.callTemplate(Xslt30Transformer.java:511)
	at org.example.Main.main(Main.java:34)
Caused by: net.sf.saxon.trans.XPathException: org.nineml.coffeegrinder.exceptions.TreeWalkerException: Unknown error: T005
	at org.nineml.coffeesacks.CommonDefinition$InvisibleXmlParserFunction.call(CommonDefinition.java:523)
	at net.sf.saxon.functions.CallableFunction.call(CallableFunction.java:128)
	at net.sf.saxon.expr.DynamicFunctionCall$DynamicFunctionCallElaborator.lambda$elaborateForPull$2(DynamicFunctionCall.java:359)
	at net.sf.saxon.expr.elab.PullElaborator.lambda$elaborateForPush$0(PullElaborator.java:38)
	at net.sf.saxon.expr.instruct.FixedElement$FixedElementElaborator.lambda$elaborateForPush$0(FixedElement.java:640)
	at net.sf.saxon.expr.instruct.NamedTemplate.expand(NamedTemplate.java:247)
	at net.sf.saxon.trans.XsltController.callTemplate(XsltController.java:872)
	at net.sf.saxon.s9api.Xslt30Transformer.callTemplate(Xslt30Transformer.java:503)
	... 1 more
Caused by: org.nineml.coffeegrinder.exceptions.TreeWalkerException: Unknown error: T005
	at org.nineml.coffeegrinder.exceptions.TreeWalkerException.getException(TreeWalkerException.java:30)
	at org.nineml.coffeegrinder.exceptions.TreeWalkerException.noTrees(TreeWalkerException.java:64)
	at org.nineml.coffeegrinder.trees.Lumberjack.getTree(Lumberjack.java:72)
	at org.nineml.coffeesacks.CommonDefinition$InvisibleXmlParserFunction.call(CommonDefinition.java:502)
	... 8 more
```

Using the latest versions of Saxon 12 and Coffeesacks:
```
    <dependencies>
        <dependency>
            <groupId>net.sf.saxon</groupId>
            <artifactId>Saxon-HE</artifactId>
            <version>12.5</version>
        </dependency>
        <dependency>
            <groupId>org.nineml</groupId>
            <artifactId>coffeesacks</artifactId>
            <version>3.2.7</version>
        </dependency>
    </dependencies>
```

John Lumley's ixml implementation jwiXML gives me a clean error
```
<ixml xmlns:ixml="http://invisiblexml.org/NS" ixml:state="failed">

End of input reached before parse completed.
Expecting one of the following terminals:
    digit:"0";
    digit:"1";
    digit:"2";
    digit:"3";
    digit:"4";
    digit:"5";
    digit:"6";
    digit:"7";
    digit:"8";
    digit:"9"
</ixml>
```

--- a/coffeesacks/src/main/java/org/nineml/coffeesacks/CoffeeSacksException.java
+++ b/coffeesacks/src/main/java/org/nineml/coffeesacks/CoffeeSacksException.java
@@ -19,6 +19,7 @@ public class CoffeeSacksException extends XPathException {
     public static final String ERR_BAD_OPTIONS = "CSER0002";
     public static final String ERR_BAD_INPUT_FORMAT = "CSER0003";
     public static final String ERR_BAD_OUTPUT_FORMAT = "CSER0004";
+    public static final String ERR_INVALID_URI = "CSER0005";
     public static final String ERR_INVALID_CHOOSE_FUNCTION = "CSIF0001";
     public static final String ERR_INVALID_CHOICE = "CSIF0002";
     public static final String ERR_TREE_CONSTRUCTION = "CSIN0001";

--- a/coffeesacks/src/main/java/org/nineml/coffeesacks/CommonDefinition.java
+++ b/coffeesacks/src/main/java/org/nineml/coffeesacks/CommonDefinition.java
@@ -22,10 +22,7 @@ import net.sf.saxon.value.AnyURIValue;
 import net.sf.saxon.value.AtomicValue;
 import net.sf.saxon.value.SequenceType;
 import net.sf.saxon.value.StringValue;
-import org.nineml.coffeefilter.InvisibleXml;
-import org.nineml.coffeefilter.InvisibleXmlDocument;
-import org.nineml.coffeefilter.InvisibleXmlParser;
-import org.nineml.coffeefilter.ParserOptions;
+import org.nineml.coffeefilter.*;
 import org.nineml.coffeefilter.trees.DataTree;
 import org.nineml.coffeefilter.trees.DataTreeBuilder;
 import org.nineml.coffeefilter.trees.SimpleTree;
@@ -41,6 +38,7 @@ import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLConnection;
 import java.nio.charset.StandardCharsets;
@@ -110,7 +108,12 @@ public abstract class CommonDefinition extends ExtensionFunctionDefinition {
                     throw new CoffeeSacksException(CoffeeSacksException.ERR_HTTP_LOOP, "Redirect limit exceeded", sourceLoc, new AnyURIValue(grammarURI.toString()));
                 }
                 seenLocations.add(location);
-                URL loc = new URL(location);
+                final URL loc;
+                try {
+                    loc = new URI(location).toURL();
+                } catch (URISyntaxException ex) {
+                    throw new CoffeeSacksException(CoffeeSacksException.ERR_INVALID_URI, ex.getMessage());
+                }
                 conn = loc.openConnection();
                 location = conn.getHeaderField("location");
             }
@@ -484,6 +487,12 @@ public abstract class CommonDefinition extends ExtensionFunctionDefinition {
                 Processor processor = (Processor) context.getConfiguration().getProcessor();
                 InvisibleXmlDocument document = parser.parse(input);
 
+                if (document instanceof InvisibleXmlFailureDocument) {
+                    InvisibleXmlFailureDocument failure = (InvisibleXmlFailureDocument) document;
+                    String message = failure.getTree();
+                    throw new XPathException(message);
+                }
+
                 final XmlForest forest;
                 try {
                     forest = new XmlForest(processor, document);
@@ -519,6 +528,8 @@ public abstract class CommonDefinition extends ExtensionFunctionDefinition {
                 }
 
                 return jsonToXDM(processor, json);
+            } catch (XPathException ex) {
+                throw ex;
             } catch (Exception ex) {
                 throw new XPathException(ex);
             }

--- a/coffeesacks/src/test/java/org/nineml/coffeesacks/StylesheetTests.java
+++ b/coffeesacks/src/test/java/org/nineml/coffeesacks/StylesheetTests.java
@@ -1,6 +1,7 @@
 package org.nineml.coffeesacks;
 
 import net.sf.saxon.s9api.*;
+import net.sf.saxon.trans.XPathException;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.nineml.coffeefilter.InvisibleXml;
@@ -19,6 +20,18 @@ public class StylesheetTests extends TestConfiguration {
         XdmNode result = transform(stylesheet, stylesheet);
         Assertions.assertEquals("<doc><date><day>15</day><month>February</month><year>2022</year></date></doc>", serialize(result));
     }
+
+    @Test
+    public void stringInputError() {
+        XdmNode stylesheet = loadStylesheet("src/test/resources/date-input-error.xsl");
+        try {
+            transform(stylesheet, stylesheet);
+            fail();
+        } catch (RuntimeException ex) {
+            Assertions.assertTrue(ex.getMessage() != null && ex.getMessage().contains("<failed"));
+        }
+    }
+
 
     @Test
     public void uriInputXmlOutput() {

--- a/coffeesacks/src/test/resources/date-input-error.xsl
+++ b/coffeesacks/src/test/resources/date-input-error.xsl
@@ -1,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
+                xmlns:xs="http://www.w3.org/2001/XMLSchema"
+                xmlns:cs="http://nineml.com/ns/coffeesacks"
+                xmlns:map="http://www.w3.org/2005/xpath-functions/map"
+                exclude-result-prefixes="#all"
+                version="3.0">
+
+<xsl:output method="xml" encoding="utf-8" indent="no"/>
+
+<xsl:mode on-no-match="shallow-copy"/>
+
+<xsl:template match="/">
+  <xsl:variable name="parser" select="cs:load-grammar(xs:anyURI('date.ixml'))"/>
+  <doc>
+    <xsl:sequence select="$parser('05 February 225')"/>
+  </doc>
+</xsl:template>
+
+</xsl:stylesheet>

